fix(anthropic): support native URL format for image inputs instead of base64 conversion

colesmcintosh opened the pull request:
## Title

fix(anthropic): support native URL format for image inputs instead of base64 conversion

## Relevant issues

Fixes #12899

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
```
============================= test session starts ==============================
tests/llm_translation/test_prompt_factory.py::test_anthropic_image_url_backward_compatibility PASSED [  9%]
tests/llm_translation/test_prompt_factory.py::test_base64_image_input[data:application/pdf;base64,1234-application/pdf] PASSED [ 18%]
tests/llm_translation/test_prompt_factory.py::test_base64_image_input[data:image/jpeg;base64,1234-image/jpeg] PASSED [ 27%]
tests/llm_translation/test_prompt_factory.py::test_base64_image_input[data:image\\/jpeg;base64,1234-image/jpeg] PASSED [ 36%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[http://example.com/test-image.jpeg-image/jpeg] PASSED [ 45%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[https://example.com/api/image-image/jpeg] PASSED [ 54%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[https://example.com/test-image.gif-image/gif] PASSED [ 63%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[https://example.com/test-image.jpg-image/jpeg] PASSED [ 72%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[https://example.com/test-image.png-image/png] PASSED [ 81%]
tests/llm_translation/test_prompt_factory.py::test_url_image_input[https://example.com/test-image.webp-image/webp] PASSED [ 90%]
tests/llm_translation/test_prompt_factory.py::test_url_image_with_format_override PASSED [100%]

====================== 11 passed, 25 deselected in 0.67s =======================
```
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🐛 Bug Fix

## Changes

### Problem
Previously, LiteLLM was converting all image URLs to base64 format before sending to Anthropic, even though Anthropic's API natively supports image URLs. This caused:
- Unnecessary base64 conversion overhead
- Larger payload sizes (base64 is ~33% larger than binary)
- Earlier hitting of Anthropic's 5MB image limit (~4MB actual image size)

### Solution
- Modified `convert_to_anthropic_image_obj()` to preserve HTTP/HTTPS URLs as native URL format
- Added support for Anthropic's native URL image type in type definitions
- Maintained full backward compatibility with base64 data URLs
- Added comprehensive test coverage for both URL and base64 formats

### Technical Details
- **URL handling**: HTTP/HTTPS URLs now use `type: "url"` instead of being converted to base64
- **Media type inference**: Automatically detects media type from URL extension or uses format parameter
- **Backward compatibility**: Base64 data URLs (`data:image/...;base64,...`) continue to work as before
- **Type safety**: Updated `AnthropicMessagesImageParam` to include `AnthropicContentParamSourceUrl`

### Files Modified
- `litellm/litellm_core_utils/prompt_templates/factory.py`: Core conversion logic
- `litellm/types/llms/anthropic.py`: Type definitions for URL support
- `tests/llm_translation/test_prompt_factory.py`: Comprehensive test coverage

### Performance Impact
- Eliminates unnecessary base64 conversion for URL-based images
- Reduces payload size by ~25% for image requests
- Allows larger images (up to 5MB actual size vs ~4MB with base64)

colesmcintosh committed:
fix(anthropic): support native URL format for image inputs instead of base64 conversion

- Add URL type support to Anthropic image objects for HTTP/HTTPS URLs
- Preserve original URLs instead of converting to base64 for better performance
- Maintain backward compatibility with base64 data URLs
- Add comprehensive test coverage for both URL and base64 formats

colesmcintosh committed:
fix(mypy): handle None return from mimetypes.guess_type in anthropic image conversion

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/anthropic-image-url-base64

krrishdholakia commented on litellm/litellm_core_utils/prompt_templates/factory.py:
```diff
@@ -727,19 +727,43 @@ def convert_to_anthropic_image_obj(
+            
+            return GenericImageParsingChunk(
+                type="url",
+                media_type=media_type,
```
this is wrong. there is no media type in the anthropic url chunk

see the doc - https://docs.anthropic.com/en/docs/build-with-claude/vision#url-based-image-example

```
{
            "type": "image",
            "source": {
              "type": "url",
              "url": "https://upload.wikimedia.org/wikipedia/commons/a/a7/Camponotus_flavomarginatus_ant.jpg"
            }
          },
```

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/anthropic-image-url-base64

colesmcintosh committed:
fix(anthropic): correct URL image format to match Anthropic API spec

- Remove media_type field from URL images per Anthropic documentation
- URL images now use {"type": "url", "url": "..."} format
- Maintain backward compatibility for base64 images with media_type
- Update GenericImageParsingChunk type to make media_type optional
- Fix tests to reflect correct Anthropic URL image specification

Resolves PR feedback about incorrect media_type in URL image chunks.

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/anthropic-image-url-base64

colesmcintosh committed:
fix(anthropic): use correct TypedDict for URL vs base64 image sources

- Use AnthropicContentParamSourceUrl for URL images instead of AnthropicContentParamSource
- Fix Vertex AI compatibility by handling both URL and base64 image types in _process_gemini_image
- Resolves MyPy type errors: missing keys, extra keys, and type incompatibility

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/anthropic-image-url-base64

colesmcintosh committed:
fix(vertex): handle both URL and base64 image types in _process_gemini_image

- Add proper type checking for convert_to_anthropic_image_obj return value
- Convert URL images to base64 for Vertex AI inline_data requirement
- Update test mocks to cover both factory and image_handling module imports
- Ensures compatibility with new URL vs base64 image handling in factory.py

jorgegarciarey commented:
@krrishdholakia , @colesmcintosh , @ishaan-berri , hi team!

Is there any inconvenience with merging this pull requests?
